### PR TITLE
Implement recursive fmt::Debug for patrica_trie::TrieDB to print Node trees

### LINKED ISSUE #8184: use debug_struct to print trie
discussion: https://github.com/paritytech/parity/pull/8166#discussion_r175972677

### PR DESCRIPTION
**Pull request by gnunicorn**
fixes #8184

### CODE REVIEW

**Comment by gnunicorn** [comment]
ah. I didn't see this part. I had run `cargo fmt` in between and it had changed... _a lot_ . I can revert that part, too, if you want  @dvdplm . I don't really have a preference...

File: util/patricia_trie/src/node.rs [hunk 1]
```diff
@@ -107,13 +108,43 @@ impl<'a> Node<'a> {
...
+				let item = item.fmt(f)?;
+				f.debug_struct("Node::Extension")
+					.field("slice", &slice)
+					.field("item", &item)
+					.finish()
+			}
+			Node::Branch(ref nodes, ref value) => {
+				let nodes_fmt = f.debug_list().entries(nodes).finish()?;
+				f.debug_struct("Node::Branch")
+					.field("nodes", &nodes_fmt)
```
Comment (debris) [inline util/patricia_trie/src/node.rs]: I think this is invalid, cause `nodes_fmt` is `()`.? [docs](https://doc.rust-lang.org/std/fmt/struct.DebugList.html#method.finish)

File: util/patricia_trie/src/node.rs [hunk 2]
```diff
@@ -107,13 +108,43 @@ impl<'a> Node<'a> {
...
+		match &self {
+			Node::Leaf(slice, value) => f.debug_struct("Node::Leaf")
+				.field("slice", &slice)
+				.field("value", &value.pretty())
+				.finish(),
+			Node::Extension(ref slice, ref item) => {
+				let item = item.fmt(f)?;
+				f.debug_struct("Node::Extension")
+					.field("slice", &slice)
+					.field("item", &item)
```
Comment (debris) [inline util/patricia_trie/src/node.rs]: I think this is invalid, cause `item` is `()`.

File: util/patricia_trie/src/node.rs [hunk 3]
```diff
@@ -107,13 +108,43 @@ impl<'a> Node<'a> {
...
 	}
 }
 
+impl<'a> fmt::Debug for Node<'a> {
+	/// Recursion helper for implementation of formatting trait.
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		match &self {
+			Node::Leaf(slice, value) => f.debug_struct("Node::Leaf")
+				.field("slice", &slice)
+				.field("value", &value.pretty())
```
Comment (debris) [inline util/patricia_trie/src/node.rs]: is it possible to avoid calling `.pretty()`? :)

Comment (gnunicorn) [reply util/patricia_trie/src/node.rs]: Yes, but that gives a different value: `00·01` (on pretty) vs `[0, 1]`. I am fine with the latter, but just to make sure that is okay.

**Comment by gnunicorn** [comment]
@debris just come to realize that you are right, for once because of the reasons you describe (formatter isn't really meant to be used recursively and the way I thought it worked was ... unsequential), but more importantly: the errors that were previously possible in the formatter came from the lookup. My current approach isn't actually doing the look up but just prints the addresses. Thus we are not seeing any tree actually.

Let me get back to this.

File: util/patricia_trie/src/node.rs [hunk 4]
```diff
@@ -177,3 +171,142 @@ impl<'a> From<Node<'a>> for OwnedNode {
...
+	use super::Node;
+	use nibbleslice::NibbleSlice;
+	static D: &'static [u8;3] = &[0x01u8, 0x23, 0x45];
+
+	#[test]
+	fn test_node_default_debug() {
+		let empty = Node::Empty;
+
+		assert_eq!(format!("{:?}", empty), "Node::Empty");
+        assert_eq!(format!("{:#?}", empty), "Node::Empty");
```
Comment (debris) [inline util/patricia_trie/src/node.rs]: bad indentation ;)

**Comment by gnunicorn** [comment]
@debris yeah. I tried to push it into Node first (handing over a closure) and that might have worked but wouldn't actually take into account the right ident.

So, I moved the entire implementation into its own `TrieAwareDebugNode` that handles lookup and all nicely and added a test on the triedb-side of things.

But as I had to remove all I did before, I decided to rebase it entirely. I hope that is okay.

File: util/patricia_trie/src/triedb.rs [hunk 5]
```diff
@@ -174,15 +119,55 @@ impl<'db> Trie for TrieDB<'db> {
...
+				Ok(Node::Leaf(slice, value)) => f.debug_struct("Node::Leaf")
+						.field("slice", &slice)
+						.field("value", &value)
+					.finish(),
+				Ok(Node::Extension(ref slice, ref item)) => f.debug_struct("Node::Extension")
+						.field("slice", &slice)
+						.field("item", &TrieAwareDebugNode{trie: self.trie, key: item})
+					.finish(),
+				Ok(Node::Branch(ref nodes, ref value)) => f.debug_struct("Node::Branch")
+						.field("nodes", &nodes)
```
Comment (gnunicorn) [inline util/patricia_trie/src/triedb.rs]: hold on a second, this still doesn't convert the entries.

**Pull request merged**